examples/pytorch/graphsage/graphsage.py

```diff
@@ -0,0 +1,295 @@
+        degs = self.g.in_degrees(nodes.nodes()).float()
+        if h.is_cuda:
+            degs = degs.cuda(h.device)
+        concatenate = torch.cat((nei, h.unsqueeze(1)), 1)
+        concatenate = torch.sum(concatenate, 1) / degs.unsqueeze(1)
+        return concatenate
+
+
+class PoolingAggregator(Aggregator):
+    def __init__(self, g, in_feats, out_feats, activation, bias):  # (2F, F)
+        super(PoolingAggregator, self).__init__(g, in_feats*2, out_feats, activation, bias)
+        self.mlp = PoolingAggregator.MLP(in_feats, in_feats, F.relu, False, True)
+
+    def concat(self, h, nei, nodes):
+        nei = self.mlp(nei)
+        concatenate = torch.cat((nei, h), 1)
+        return concatenate
+
+    class MLP(nn.Module):
+        def __init__(self, in_feats, out_feats, activation, dropout, bias):  # (F, F)
+            super(PoolingAggregator.MLP, self).__init__()
+            self.weight = nn.Parameter(torch.Tensor(in_feats, out_feats))  # (F,EF)
+            if dropout:
+                self.dropout = nn.Dropout(p=dropout)
+            else:
+                self.dropout = 0.
+            if bias:
+                self.bias = nn.Parameter(torch.Tensor(out_feats))  # (EF)
+            else:
+                self.bias = None
+            self.activation = activation
+            self.reset_parameters()
+
+        def reset_parameters(self):
+            weight_stdv = 1. / math.sqrt(self.weight.size(1))
+            self.weight.data.uniform_(-weight_stdv, weight_stdv)
+            if self.bias is not None:
+                bias_stdv = 1. / math.sqrt(self.bias.size(0))
+                self.bias.data.uniform_(-bias_stdv, bias_stdv)
+
+        def forward(self, nei):  # (B, N, F)
+            if self.dropout:
+                nei = self.dropout(nei)
+            nei = torch.matmul(nei, self.weight)  # (B, N, EF)
+            if self.bias is not None:
+                nei = nei + self.bias
+            if self.activation:
+                nei = self.activation(nei)
+            max_value = torch.max(nei, dim=1)[0]  # (B, EF)
+            return max_value
+
+
+class GraphSAGELayer(nn.Module):
+    def __init__(self,
+                 g,
+                 in_feats,
+                 out_feats,
+                 activation,
+                 dropout,
+                 aggregator_type,
+                 bias=True,
+                 ):
+        super(GraphSAGELayer, self).__init__()
+        self.g = g
+        if dropout:
+            self.dropout = nn.Dropout(p=dropout)
+        else:
+            self.dropout = 0.
+        if aggregator_type == "pooling":
+            self.aggregator = PoolingAggregator(g, in_feats, out_feats, activation, bias)
+        else:
+            self.aggregator = MeanAggregator(g, in_feats, out_feats, activation, bias)
+
+    def forward(self, h):
+        if self.dropout:
+            h = self.dropout(h)
+        self.g.ndata['h'] = h
+        self.g.update_all(graphsage_msg, self.aggregator)
```

Comment:
You can use the built-in message function `fn.copy_src(src='h', out='m')`

Reply:
Thanks for your reivew! I got it!

Reply:
Meaning: replace `graphsage_msg` with `fn.copy_src(src='h', out='m')`.

Reply:
What is the difference between them?

Reply:
There is no difference semantically.  However,
* You write one less Python function,
* It allows possibility for DGL to further optimize on speed, since you are using a DGL-provided builtin function.